Add Hash and RoundRobinPartitionFunctionSpecs

kewang1024 opened the pull request:
HashPartitionFunctionSpec and RoundRobinPartitionFunctionSpec

mbasmanova commented on velox/exec/tests/RoundRobinPartitionFunctionTest.cpp:
```diff
@@ -37,3 +42,11 @@ TEST(RoundRobinPartitionFunctionTest, basic) {
 }
+
+TEST_F(RoundRobinPartitionFunctionTest, spec) {
+  auto roundRobinPartitionFunctionSpec =
```
roundRobinPartitionFunctionSpec -> spec for readability (or roundRobinSpec)

mbasmanova commented on velox/exec/tests/PlanNodeToStringTest.cpp:
```diff
@@ -492,7 +492,7 @@ TEST_F(PlanNodeToStringTest, localPartition) {
   ASSERT_EQ("-- LocalPartition\n", plan->toString());
   ASSERT_EQ(
-      "-- LocalPartition[REPARTITION HASH(0)] -> c0:SMALLINT, c1:INTEGER, c2:BIGINT\n",
+      "-- LocalPartition[REPARTITION HASH(keys:0, constValues:)] -> c0:SMALLINT, c1:INTEGER, c2:BIGINT\n",
```
Can we avoid printing "constValues:" is there are no constants?

mbasmanova commented on velox/exec/tests/HashPartitionFunctionTest.cpp:
```diff
@@ -114,3 +114,19 @@ TEST_F(HashPartitionFunctionTest, HashPartitionFunction) {
+  RowTypePtr outputType(
+      ROW({"c0", "c1", "c2", "c3", "c4"},
+          {BIGINT(), SMALLINT(), INTEGER(), BIGINT(), VARCHAR()}));
+  core::PartitionFunctionSpecPtr hashPartitionFunctionSpec =
```
hashPartitionFunctionSpec -> auto spec for readability

mbasmanova commented on velox/exec/PartitionFunction.h:
```diff
@@ -0,0 +1,35 @@
+
+namespace facebook::velox::exec {
+
+static void registerPartitionFunctionSerDe() {
```
No need to make this 'static'. Let's also move the implementation to .cpp file.

mbasmanova commented on velox/exec/PartitionFunction.h:
```diff
@@ -0,0 +1,35 @@
+#pragma once
+#include <velox/exec/HashPartitionFunction.h>
+#include <velox/exec/RoundRobinPartitionFunction.h>
+#include "velox/core/PlanNode.h"
```
Is this needed?

mbasmanova commented on velox/exec/HashPartitionFunction.cpp:
```diff
@@ -84,4 +84,62 @@ void HashPartitionFunction::partition(
+  std::vector<velox::core::ConstantTypedExpr> constValues;
+  constValues.reserve(constValues_.size());
+  for (const auto& value : constValues_) {
+    constValues.emplace_back(value);
```
Can 'value' be null? Do we have a test for this case?

mbasmanova commented:
CC: @kgpai Krishna, how far along are you with using plan serde in the fuzzers? Will this change affect your work?

xiaoxmeng commented on velox/exec/PartitionFunction.cpp:
```diff
@@ -0,0 +1,36 @@
+  registry.Register(
+      "RoundRobinPartitionFunctionSpec",
+      RoundRobinPartitionFunctionSpec::deserialize);
+  registry.Register(
```
Do we need to remove the register in PlanNode::registerSerDe()? @mbasmanova Thanks!

mbasmanova replied:
PlanNode::registerSerDe registers GatherPartitionFunctionSpec, hence, we cannot just remove it. We can move GatherPartitionFunctionSpec out of PlanNode.h into its own file like RoundRobin and Hash specs, then add serde here.

xiaoxmeng replied:
GatherPartitionFunctionSpec still need to stay in core namespace as it is directly used by PlanNode.h/cpp. Thanks!

xiaoxmeng commented on velox/exec/HashPartitionFunction.h:
```diff
@@ -58,4 +58,31 @@ class HashPartitionFunction : public core::PartitionFunction {
+      RowTypePtr inputType,
+      std::vector<column_index_t> keys,
+      std::vector<VectorPtr> constValues)
+      : inputType_{inputType},
```
```
inputType_{std::move(inputType)},
```

xiaoxmeng commented on velox/exec/HashPartitionFunction.cpp:
```diff
@@ -84,4 +84,67 @@ void HashPartitionFunction::partition(
+}
+
+std::string HashPartitionFunctionSpec::toString() const {
+  std::ostringstream constValues;
```
How about
```
std::vector<std::string> constValueStrs;
constValueStrs.reserve(constValues_.size());
for (const auto& value : constValues_) {
  constValueStrs.emplace_back(value->toString());
}
return constValueStrs.empty() ? fmt::format("HASH(KEYS:{})", folly::join(", ", keys_)) : fmt::format(
            "HASH(KEYS:{}, CONSTS:{})",
            folly::join(", ", keys_),
            folly::join(", ", constValueStrs)));
```
s/keys/KEYS/
s/constValues/CONSTS/

xiaoxmeng commented on velox/exec/HashPartitionFunction.cpp:
```diff
@@ -84,4 +84,67 @@ void HashPartitionFunction::partition(
+  std::vector<velox::core::ConstantTypedExpr> constValues;
+  constValues.reserve(constValues_.size());
+  for (const auto& value : constValues_) {
+    if (value) {
```
value must not be null? 
```
for (const auto& value : constValues_) {
  VELOX_CHECK_NOT_NULL(value);
  ...
}
```

xiaoxmeng commented on velox/exec/tests/HashPartitionFunctionTest.cpp:
```diff
@@ -114,3 +114,36 @@ TEST_F(HashPartitionFunctionTest, HashPartitionFunction) {
+  auto hashSpec = std::make_unique<exec::HashPartitionFunctionSpec>(
+      outputType,
+      std::vector<column_index_t>{0, 1},
+      std::vector<VectorPtr>{nullptr});
```
Can this be null? Shall we test std::vector<VectorPtr> is empty instead?

mbasmanova replied:
I imagine that there can be 5 keys, some of which are constant and some not. For example, first 2 keys are constant or last 2 keys are constant or every other key is constant or none of the keys are constant. Let's make sure to test all these cases.

xiaoxmeng reviewed (commented):
@kewang1024 thanks for the split!

kgpai commented:
@mbasmanova @kewang1024 I too have some changes in PlanBuilder, however i can merge these with Ke's changes if her PR lands before mine.

mbasmanova commented on velox/exec/HashPartitionFunction.cpp:
```diff
@@ -84,4 +84,67 @@ void HashPartitionFunction::partition(
+      : fmt::format(
+            "HASH(keys:{}, constValues:{})",
+            folly::join(", ", keys_),
+            constValues.str());
```
use constValuesStr variable

mbasmanova commented on velox/exec/HashPartitionFunction.h:
```diff
@@ -58,4 +58,31 @@ class HashPartitionFunction : public core::PartitionFunction {
+
+class HashPartitionFunctionSpec : public core::PartitionFunctionSpec {
+ public:
+  HashPartitionFunctionSpec(
```
Let's add some documentation to explain the relationship between constValues and keys. I assume constValues.size must be equal to keys.size and some values can be null, while values that correspond to kConstantChannel must be not null. Let's also add sanity checks to enforce this.

mbasmanova commented on velox/exec/HashPartitionFunction.h:
```diff
@@ -58,4 +58,31 @@ class HashPartitionFunction : public core::PartitionFunction {
+  HashPartitionFunctionSpec(
+      RowTypePtr inputType,
+      std::vector<column_index_t> keys,
+      std::vector<VectorPtr> constValues)
```
do not abbreviate and use keys instead of values (these are values of the keys, but using just values is confusing)

constantKeys

mbasmanova commented on velox/exec/tests/HashPartitionFunctionTest.cpp:
```diff
@@ -114,3 +114,36 @@ TEST_F(HashPartitionFunctionTest, HashPartitionFunction) {
+
+TEST_F(HashPartitionFunctionTest, spec) {
+  Type::registerSerDe();
+  auto& registry = DeserializationWithContextRegistryForSharedPtr();
```
Use ITypedExpr::registerSerDe

mbasmanova commented on velox/exec/tests/HashPartitionFunctionTest.cpp:
```diff
@@ -114,3 +114,36 @@ TEST_F(HashPartitionFunctionTest, HashPartitionFunction) {
+
+  auto serialized = hashSpec->serialize();
+  auto copy = HashPartitionFunctionSpec::deserialize(serialized, pool());
+  ASSERT_EQ(serialized["constants"].size(), 0);
```
Let's assert that copy and hashSpec are the same, for example, by comparing the output of toString() method.

mbasmanova reviewed (commented):
@kewang1024 Ke, thank you for iterating on this PR. Some follow-up comments.

kewang1024 commented:
Thanks @mbasmanova for your diligent review!

Regarding naming and size check of `constValues` and `keys` for `HashPartitionFunctionSpec`, I don't think we need to change the current logic, from `HashPartitionFunction`, we can see that 
1. keys refers to the non-constant keys we want to partition on (see `keyChannels` in `HashPartitionFunction`)
2. constVales refers to the constant key

I'm not sure why initial naming is `keys`, but I just changed it to `keyChannels` in `HashPartitionFunctionSpec` as well to be more consistent

mbasmanova commented:
> Regarding naming and size check...

@kewang1024 Thank you for clarifying. Let's document this class to explain the properties.

Also, I wonder whether the order of key channels and constant values matters or the hash function used is order-agnostic. Would you clarify?

kewang1024 commented:
Hey @mbasmanova, my apologies, for the statement above, conclusion is right but the reason is slightly wrong, here is the details:
1. 'keyChannels_' stores the index of keys to partition on, if the key is a constant, use index 'kConstantChannel' to indicate so and store the constant value as a base vector in 'constValues_'.
2. The 'constValues_' size is less than or equal to 'keyChannels_' size.
3. There wouldn't be null in 'constValues_' vector

So based on those, I removed our `null` test and also added the class description



Regarding the order of key channels and constant values, I didn't put any comment there saying it's order-agnostic or not because it might introduce some confusion; 
1. The order of them would affect the *Hash value* result according to implementation in `HashPartitionFunction`, hash(A, B) != hash(B, A)
2. But records w/ the same A, B value will go to the same partition , in this sense, (A, B) is same as (B, A)

mbasmanova commented:
@kewang1024 Ke, thank you for clarifications.

> 1. 'keyChannels_' stores the index of keys to partition on, if the key is a constant, use index 'kConstantChannel' to indicate so and store the constant value as a base vector in 'constValues_'.
> 2. The 'constValues_' size is less than or equal to 'keyChannels_' size.
> 3. There wouldn't be null in 'constValues_' vector

Why would there be no nulls in constValues_? What if first key is a column and second is a constant?

> 1. The order of them would affect the _Hash value_ result according to implementation in `HashPartitionFunction`, hash(A, B) != hash(B, A)
> 2. But records w/ the same A, B value will go to the same partition , in this sense, (A, B) is same as (B, A)

Why would the records go to the same partition? If hash value is different, then partition number can be different as well, no?

kewang1024 commented:
> @kewang1024 Ke, thank you for clarifications.
> 
> > 1. 'keyChannels_' stores the index of keys to partition on, if the key is a constant, use index 'kConstantChannel' to indicate so and store the constant value as a base vector in 'constValues_'.
> > 2. The 'constValues_' size is less than or equal to 'keyChannels_' size.
> > 3. There wouldn't be null in 'constValues_' vector
> 
> Why would there be no nulls in constValues_? What if first key is a column and second is a constant?
> 
> > 1. The order of them would affect the _Hash value_ result according to implementation in `HashPartitionFunction`, hash(A, B) != hash(B, A)
> > 2. But records w/ the same A, B value will go to the same partition , in this sense, (A, B) is same as (B, A)
> 
> Why would the records go to the same partition? If hash value is different, then partition number can be different as well, no?



> @kewang1024 Ke, thank you for clarifications.
> 
> > 1. 'keyChannels_' stores the index of keys to partition on, if the key is a constant, use index 'kConstantChannel' to indicate so and store the constant value as a base vector in 'constValues_'.
> > 2. The 'constValues_' size is less than or equal to 'keyChannels_' size.
> > 3. There wouldn't be null in 'constValues_' vector
> 
> Why would there be no nulls in constValues_? What if first key is a column and second is a constant?
> 
> > 1. The order of them would affect the _Hash value_ result according to implementation in `HashPartitionFunction`, hash(A, B) != hash(B, A)
> > 2. But records w/ the same A, B value will go to the same partition , in this sense, (A, B) is same as (B, A)
> 
> Why would the records go to the same partition? If hash value is different, then partition number can be different as well, no?

Say you have a sql "select A, 10, B from tableX group by 1, 2, 3";
Then it would look like this: 
- the keyChannels_       (0, max_index, 2)
- constValues_              ([10])
----------------------------------------------------
Yes, the partition number would be different, what I meant was records w/ the same A, B value will still go to the same partition
For example Hash(A, B):
L1      (1, 3) -> 13
L2     (1, 4) -> 14
L3     (1, 3) -> 13
........

Then for Hash(B, A):
L1     (3, 1) -> 31
L2    (4, 1) -> 41
L3    (3, 1) -> 31


So Hash (A, B) != Hash (B, A) but for records with same value would go to the same partition, L1 and L3 would always be in the same partition regardless of the order

mbasmanova commented:
@kewang1024 

> Say you have a sql "select A, 10, B from tableX group by 1, 2, 3";
> Then it would look like this:
> the keyChannels_ (0, max_index, 2)
> constValues_ ([10])

Is this correct? I'd expect constValues_ to be [null, 10, null] or [null, 10].

kewang1024 commented:
Looking at the `HashPartitionFunction` implementation

```
void HashPartitionFunction::init(
    const RowTypePtr& inputType,
    const std::vector<column_index_t>& keyChannels,
    const std::vector<VectorPtr>& constValues) {
  hashers_.reserve(keyChannels.size());
  size_t constChannel{0};
  for (const auto channel : keyChannels) {
    if (channel != kConstantChannel) {
      hashers_.emplace_back(
          VectorHasher::create(inputType->childAt(channel), channel));
    } else {
      const auto& constValue = constValues[constChannel++];
      hashers_.emplace_back(VectorHasher::create(constValue->type(), channel));
      hashers_.back()->precompute(*constValue);
    }
  }
}
```

mbasmanova commented on velox/exec/HashPartitionFunction.h:
```diff
@@ -58,4 +58,36 @@ class HashPartitionFunction : public core::PartitionFunction {
+  HashPartitionFunctionSpec(
+      RowTypePtr inputType,
+      std::vector<column_index_t> keyChannels,
+      std::vector<VectorPtr> constValues = {})
```
@kewang1024 Do we need this default value? Let's remove it.

kewang1024 replied:
I also thought about this one, but I'm more leaning towards keeping it given:
1. This call is used in several places, and having the default value makes the code cleaner and more straighforward
2. From the default value, it's easy to infer that constValues_.size != keyChannels_.size, it's a good implications for the relationships between constValues_ and keyChannels_

Let me know if you have strong opinion against it

mbasmanova commented on velox/exec/HashPartitionFunction.h:
```diff
@@ -58,4 +58,36 @@ class HashPartitionFunction : public core::PartitionFunction {
+/// 'keyChannels_' stores the index of keys to partition on, if the key is a
+/// constant, use index 'kConstantChannel' to indicate so and store the constant
+/// value as a base vector in 'constValues_'
+/// The 'constValues_' size is less than or equal to 'keyChannels_' size
```
Use public argument names, not private member variables.

keyChannels_ -> keyChannels, etc.

mbasmanova commented on velox/exec/HashPartitionFunction.cpp:
```diff
@@ -84,4 +84,59 @@ void HashPartitionFunction::partition(
+    constValueStrs.emplace_back(value->toString());
+  }
+  return constValueStrs.empty()
+      ? fmt::format("HASH(keyChannels:{})", folly::join(", ", keyChannels_))
```
Do we have a test case for toString() output when there are constant values?

kewang1024 replied:
Yes

mbasmanova commented on velox/exec/PartitionFunction.cpp:
```diff
@@ -0,0 +1,35 @@
+      "RoundRobinPartitionFunctionSpec",
+      RoundRobinPartitionFunctionSpec::deserialize);
+  registry.Register(
+      "GatherPartitionFunctionSpec",
```
Do we register this in 2 places now? Here and in PlanNode::registerSerDe?

kewang1024 replied:
Yes, plan to remove in following PR when we change all the related spec registery
